libs/nss.c

```diff
@@ -108,8 +120,15 @@ int _nss_tcb_getspent_r(struct spwd *__result_buf,
 
 	do {
 		currpos = telldir(tcbdir);
+#ifdef GLIBC_2_24
+		saved_errno = errno;
+		errno = 0;
+		result = readdir(tcbdir);
+		if (errno != 0) {
```

Comment:
`errno` is defined if and only  if `result == NULL`.
Also, `saved_errno` is not set.

Reply:
`saved_errno` is set in Line 124.

Reply:
OK, `saved_errno` is set, but `errno` is undefined unless `result == NULL`.

Reply:
I'd like to highlight: if `errno` is reset to `0` before `readdir`, then `errno != 0` after `readdir` indicates an error if and only if `readdir` returned `NULL`. If `readdir` returned nonzero, then the value of `errno` is undefined.

Reply:
Well, yes, `errno == 0` until a failure in `readdir` will occur, which additionally will result in `result == NULL`.

I assume, you want the code to explicitly check for `(errno && !result)`?

Reply:
> I'd like to highlight: if `errno` is reset to `0` before `readdir`, then `errno != 0` after `readdir` indicates an error if and only if `readdir` returned `NULL`. If `readdir` returned nonzero, then the value of `errno` is undefined.

Ok, now I got you.

Reply:
Yes, I mean `(!result && errno)`.

Reply:
Fixed in rebased commit.